addons/l10n_in_ewaybill_stock/models/l10n_in_ewaybill.py

```diff
@@ -0,0 +1,603 @@
+        if not self.handle_error_if_present(response):
+            self._write_successfully_response({'state': 'cancel'})
+
+    def _generate_ewaybill_direct(self):
+        ewb_api = EWayBillApi(self.company_id)
+        generate_json = self._ewaybill_generate_direct_json()
+        response = ewb_api._ewaybill_generate(generate_json)
+        if not self.handle_error_if_present(response):
+            self._handle_internal_warning_if_present(response)  # In case of error 604
+            self.state = 'generated'
+            response_data = response.get("data")
+            self._write_successfully_response({
+                'name': response_data.get("ewayBillNo"),
+                'ewaybill_date': self._indian_timezone_to_odoo_utc(
+                    response_data.get('ewayBillDate')
+                ),
+                'ewaybill_expiry_date': self._indian_timezone_to_odoo_utc(
+                    response_data.get('validUpto')
+                ),
+            })
+
+    @api.model
+    def _indian_timezone_to_odoo_utc(self, str_date, time_format='%d/%m/%Y %I:%M:%S %p'):
+        """
+            This method is used to convert date from Indian timezone to UTC
+        """
+        try:
+            local_time = datetime.strptime(str_date, time_format)
+        except ValueError:
+            try:
+                local_time = datetime.strptime(str_date, "%d/%m/%Y %H:%M:%S ")
+            except ValueError:
+                # Worst senario no date format matched
+                _logger.warning("Something went wrong while L10nInEwaybill date conversion")
+                return fields.Datetime.to_string(fields.Datetime.now())
+        utc_time = local_time.astimezone(pytz.utc)
+        return fields.Datetime.to_string(utc_time)
+
+    @api.model
+    def _get_partner_state_code(self, partner):
+        return int(partner.state_id.l10n_in_tin) if partner.country_id.code == "IN" else 99
+
+    def _l10n_in_tax_details(self):
+        tax_details = {
+            'line_tax_details': defaultdict(dict),
+            'tax_details': defaultdict(float)
+        }
+        for move in self.move_ids:
+            line_tax_vals = self._l10n_in_tax_details_by_line(move)
+            tax_details['line_tax_details'][move.id] = line_tax_vals
+            for val_field in ['total_excluded', 'total_included', 'total_void']:
+                tax_details['tax_details'][val_field] += line_tax_vals[val_field]
+            for tax in ['igst', 'cgst', 'sgst', 'cess_non_advol', 'cess', 'other']:
+                for taxes in line_tax_vals['taxes']:
+                    for field_key in ["rate", "amount"]:
+                        if (key := f"{tax}_{field_key}") in taxes:
+                            tax_details['tax_details'][key] += taxes[key]
+        return tax_details
+
+    def _l10n_in_tax_details_by_line(self, move):
+        taxes = move.ewaybill_tax_ids.compute_all(price_unit=move.ewaybill_price_unit, quantity=move.quantity)
```

Comment:
We can not have a discount in sale/purchase?

Reply:
@hamo-odoo This seems ok as the ewaybill_price_unit is calculated from the purchase order line subtotal e.g.